### PR TITLE
Add `CObjectPool`, the replacement of `alloc.h` `new` overriding macros

### PR DESCRIPTION
**Pull request by gerdoe-jr**
There's nothing special, switched from previous global pool to local one. Next step is unifying `m_apPlayers` calls by introducing new `CObjectPool *PlayerPool() const` or `CPlayer *GetPlayer(int ClientID) const`.

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/game/server/alloc.h [hunk 1]
```diff
@@ -3,78 +3,86 @@
...
-#define MACRO_ALLOC_GET_SIZE(POOLTYPE) (sizeof(POOLTYPE))
-#endif
+template<typename Type, const std::size_t Size>
+class CObjectPool {
+    char m_Data[Size][sizeof(Type)]{{0}};
+    std::bitset<Size> m_Used{0};
+
+public:
+    CObjectPool() {
+    }
```
Comment (Chairn) [inline src/game/server/alloc.h:18]: You don't need constructor given you use c++11 assignments in member variable declarations

File: src/game/server/alloc.h [hunk 2]
```diff
@@ -3,78 +3,86 @@
...
-private:
-
-#if __has_feature(address_sanitizer)
-#define MACRO_ALLOC_GET_SIZE(POOLTYPE) ((sizeof(POOLTYPE) + 7) & ~7)
-#else
-#define MACRO_ALLOC_GET_SIZE(POOLTYPE) (sizeof(POOLTYPE))
-#endif
+template<typename Type, const std::size_t Size>
+class CObjectPool {
+    char m_Data[Size][sizeof(Type)]{{0}};
```
Comment (Chairn) [inline src/game/server/alloc.h:13]: m_aaData

**Comment by Learath2** [comment]
I'd like to see the ASan poisoning back

File: src/game/server/alloc.h [hunk 3]
```diff
@@ -3,78 +3,86 @@
...
+    std::bitset<Size> m_Used{0};
+
+public:
+    CObjectPool() {
+    }
+
+    template <typename ...TArgs>
+    Type *New(std::size_t ObjectId, TArgs ...Args) {
```
Comment (Robyt3) [inline src/game/server/alloc.h:21]: `std::size_t` -> `size_t`. We are not using the namespace for this type anywhere.

File: src/game/server/alloc.h [hunk 4]
```diff
@@ -3,78 +3,86 @@
...
+        dbg_assert(!m_Used[ObjectId], "Already constructed");
+        dbg_assert(0 <= ObjectId && ObjectId < Size, "Invalid ObjectId");
```
Comment (Robyt3) [inline src/game/server/alloc.h:23]: `0 <= ObjectId` is redundant for `size_t`, which cannot be negative.

File: src/game/server/alloc.h [hunk 5]
```diff
@@ -3,78 +3,86 @@
...
+
+        Type *Pointer = new (m_Data[ObjectId]) Type(std::forward<TArgs>(Args)...);
```
Comment (Robyt3) [inline src/game/server/alloc.h:25]: `Pointer` -> `pPointer`

File: src/game/server/gamecontext.cpp [hunk 6]
```diff
@@ -1618,19 +1616,19 @@ void CGameContext::OnClientDrop(int ClientID, const char *pReason)
 	AbortVoteKickOnDisconnect(ClientID);
 	m_pController->OnPlayerDisconnect(m_apPlayers[ClientID], pReason);
 	delete m_apPlayers[ClientID];
```
Comment (Anime-pdf) [inline src/game/server/gamecontext.cpp:1618]: Should be removed as well

File: src/game/server/alloc.h [hunk 7]
```diff
@@ -3,78 +3,90 @@
...
+		dbg_assert(m_Used[ObjectId], "Already unused");
+		dbg_assert(ObjectId < Size, "Invalid ObjectId");
+
+		m_Used[ObjectId] = false;
+	}
+
+	void Reset()
+	{
+		for(size_t i = 0; i < Size; i++)
+			m_Used[i] = false;
```
Comment (Robyt3) [inline src/game/server/alloc.h:53]: Use `m_Used.reset()` to clear the entire bitset at once.

File: src/game/server/gamecontext.h [hunk 8]
```diff
@@ -165,7 +167,8 @@ class CGameContext : public IGameServer
 	void Clear();
 
 	CEventHandler m_Events;
-	CPlayer *m_apPlayers[MAX_CLIENTS];
+	CObjectPool<CPlayer, MAX_CLIENTS> m_apPlayers;
+    CObjectPool<CCharacter, MAX_CLIENTS> m_apCharacters;
```
Comment (Robyt3) [inline src/game/server/gamecontext.h:171]: Remove the `ap` prefixes, those are no longer arrays of pointers.

**Comment by gerdoe-jr** [comment]
Also, what do you guys think, is `GetPlayer()` alias with array bounds checks okey? I want to replace all `m_apPlayers` calls with that. `GetPlayerUnchecked()`? xd

**Comment by gerdoe-jr** [comment]
Okey, there's actually no reason to put any of such alias in this PR, so it goes as it is.

**Pull request closed without merging**